packages/flutter/lib/src/material/refresh_indicator.dart

```diff
@@ -0,0 +1,239 @@
+  Animation<double> _scaleFactor;
+  Animation<Color> _valueColor;
+
+  double _scrollOffset;
+  double _containerExtent;
+  double _minScrollOffset;
+  double _maxScrollOffset;
+  RefreshIndicatorLocation _location = RefreshIndicatorLocation.top;
+
+  @override
+  void initState() {
+    super.initState();
+    _sizeFactor = new Tween<double>(begin: 0.0, end: _kDragSizeFactorLimit).animate(_sizeController);
+    _scaleFactor = new Tween<double>(begin: 1.0, end: 0.0).animate(_scaleController);
+
+    final ThemeData theme = Theme.of(context);
+
+    // Fully opaque when we've reached config.displacement.
+    _valueColor = new ColorTween(
+      begin: theme.primaryColor.withOpacity(0.0),
+      end: theme.primaryColor.withOpacity(1.0)
+    )
+    .animate(new CurvedAnimation(
+      parent: _sizeController,
+      curve: new Interval(0.0, 1.0 / _kDragSizeFactorLimit)
+    ));
+  }
+
+  void _updateState(ScrollableState scrollable) {
+    final Axis axis = scrollable.config.scrollDirection;
+    if (axis != Axis.vertical || scrollable.scrollBehavior is! ExtentScrollBehavior)
+      return;
+    final ExtentScrollBehavior scrollBehavior = scrollable.scrollBehavior;
+    _scrollOffset = scrollable.scrollOffset;
+    _containerExtent = scrollBehavior.containerExtent;
+    _minScrollOffset = scrollBehavior.minScrollOffset;
+    _maxScrollOffset = scrollBehavior.maxScrollOffset;
+  }
+
+  void _onScrollStarted(ScrollableState scrollable) {
+    _updateState(scrollable);
+    _scaleController.value = 0.0;
+    _sizeController.value = 0.0;
+  }
+
+  RefreshIndicatorLocation get _locationForScrollOffset {
+    return _scrollOffset < _minScrollOffset
+      ? RefreshIndicatorLocation.top
+      : RefreshIndicatorLocation.bottom;
+  }
+
+  void _onScrollUpdated(ScrollableState scrollable) {
+    final double value = scrollable.scrollOffset;
+    if ((value < _minScrollOffset || value > _maxScrollOffset) &&
+        ((value - _scrollOffset).abs() > kPixelScrollTolerance.distance)) {
+      final double overScroll = value < _minScrollOffset ? _minScrollOffset - value : value - _maxScrollOffset;
+      final double newValue = overScroll / (_containerExtent * _kDragContainerExtentPercentage);
+      if (newValue > _sizeController.value) {
+        _sizeController.value = newValue;
+        if (_location != _locationForScrollOffset) {
+          setState(() {
+            _location = _locationForScrollOffset;
+          });
+        }
+      }
+    }
+    _updateState(scrollable);
+  }
+
+  Future<Null> _defaultRefresh() {
+    Completer<Null> completer = new Completer<Null>();
+    new Timer(new Duration(seconds: 3), () { completer.complete(null); });
+    return completer.future;
+  }
+
+  Future<Null> _doOnScrollEnded(ScrollableState scrollable) async {
+    if (_valueColor.value.alpha == 0xFF) {
+      await _sizeController.animateTo(1.0 / _kDragSizeFactorLimit, duration: _kIndicatorSnapDuration);
+      await (config.refresh ?? _defaultRefresh)();
```

Comment:
I think config.refresh should be required. If you don't have something to refresh, then why have this component?